Updates to package file format read/write, incl. refactor, plus unit tests

mcdurdin opened the pull request:
This PR has been broken into several parts:

1. Removal of obsolete code (commit:36b7c7f)
2. Refactoring of the package file format classes to use (some) more modern classes (commit:1572616)
3. Changes to package file format, addition of unit test framework, plus test case (the rest)

darcywong00 commented on windows/src/unit-tests/package-info/test/test1.inf:
```diff
@@ -0,0 +1,62 @@
+OSKFont=wookianos.ttf
+DisplayFont=fantuwua.ttf
+Language0=amh,Amharic
+Language1=gez,Ge'ez
```
Should `Ge'ez` be in quotes `"Ge'ez"` or is that the test scenario?

jahorton commented on windows/src/global/delphi/general/PackageInfo.pas:
```diff
@@ -122,19 +122,16 @@   TPackageRegistryKey = class
     procedure SaveIni(AIni: TIniFile); virtual;
     procedure LoadXML(ARoot: IXMLNode); virtual;
     procedure SaveXML(ARoot: IXMLNode); virtual;
-    property Items[Index: Integer]: TPackageRegistryKey read Get write Put; default;
```
Are these implemented automatically by the generic base class?  I assume that's why they're being removed.

mcdurdin replied:
Correct

jahorton commented on windows/src/global/delphi/general/kmpinffile.pas:
```diff
@@ -162,6 +162,11 @@ procedure TKMPInfFile.Import(AIni: TIniFile);
+
+//    TODO
+//    Also remember to update version number if source files change
+//    Also do not permit different versions of same keyboard in compiled form!
```
Does this have an issue, if we're leaving it to a future PR?  Or was this accidentally left out?

mcdurdin replied:
This will be in a future PR. I should pull it out of the source code and put it into the trello card.

mcdurdin commented:
@darcywong00 The string is valid as it stands. The single quote in the middle is the test scenario.

mcdurdin commented:
Okay, awaiting .xml format update - don't review yet.

mcdurdin commented:
The latest commits (5d1be46-a395666) add support for JSON format and refactor a little further. This does make the PR sadly a little larger :(

darcywong00 commented on windows/src/unit-tests/package-info/.gitignore:
```diff
@@ -0,0 +1,2 @@
+test/*.out.*
+Win32
```
can we consolidate .gitignore to the root platform directory?

mcdurdin replied:
I'm going to suggest that minimal localised .gitignore are less side-effect prone and help us to see that files are ignored in the folders we are working in. When it's a top-level .gitignore, then we tend not to see it...

mcdurdin replied:
(I have, in a separate PR, consolidated the Win32 folder to a parent level folder because there are two projects that use it and it really is a common ignore, unlike the test/*.out.* filespec)

darcywong00 commented on windows/src/unit-tests/package-info/test/test.xml:
```diff
@@ -0,0 +1,2 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Package><System><KeymanDeveloperVersion>9.0.500.0</KeymanDeveloperVersion><FileVersion>7.0</FileVersion></System><Options><ExecuteProgram></ExecuteProgram><ReadMeFile>readme.htm</ReadMeFile><GraphicFile>sideimage.bmp</GraphicFile><MSIFileName></MSIFileName><MSIOptions></MSIOptions></Options><Buttons><Button><ButtonType>install</ButtonType><Caption>&amp;Install</Caption><URL></URL><Left>0</Left>< ...
```
Can you beautify this xml file so it'll be easier to maintain the unit tests and easily see diffs in future PR's?

mcdurdin replied:
Sure

darcywong00 commented on windows/src/unit-tests/package-info/test/test.json:
```diff
@@ -0,0 +1,187 @@
+  "files" : [
+    {
+      "name" : "C:\\Projects\\keyman\\open\\windows\\src\\unit-tests\\package-info\\Win32\\Debug\\..\\..\\test\\readme.htm",
+      "description" : "File readme.htm"
```
Do the tests work if someone isn't using `C:\\Projects\keyman\...` as a root path? 
I see the test.xml and test.ini only use relative paths.

mcdurdin replied:
Yeah, I'll update the test to use relative paths. Good catch.

darcywong00 commented on windows/src/unit-tests/package-info/PackageInfoTest.pas:
```diff
@@ -0,0 +1,299 @@
+
+    [Test]
+    procedure TestAssign;
+  end;
```
What are your thoughts on making another test that verifies `LoadIni`, `LoadXml`, and `LoadJSON` give you identical structures? That way, when we update the packageInfo structure, we'll have a test that Developer is keeping all 3 formats in sync.

jahorton replied:
+1

mcdurdin replied:
That's pretty well tested with the round trip tests, no?

mcdurdin commented:
@jahorton Good point. I've done a code check and `RegistryKeys` is no longer used (was used for standalone installers in the past), so it's gone. But in doing so, noted some missing tests for `StartMenu` etc which I've added in, and in doing that have also corrected some additional roundtripping issues there.
